File: python/BF.py
#!/usr/bin/env python
# encoding=UTF-8


def bf(string_S, string_T):
    """
    字符串的模式匹配
    """
    s_len = len(string_S)
    t_len = len(string_T)
    for i in range(s_len - 1):
        k = i
        j = 0
        while (string_S[k] == string_T[j]):
            k += 1
            j += 1
            if j == t_len:
                return i
    return 0


def bf_1(string_S, string_T):
    """
    字符串的模式匹配
    """
    s_len = len(string_S)
    t_len = len(string_T)
    i = 0
    j = 0
    while(i < s_len and j < t_len):
        if string_S[i] == string_T[j]:
            i += 1
            j += 1
        else:
            j = 0
            i = i - j + 1
    if j >= t_len:
        return i - t_len
    return 0


def main(st, rt):
    # pos_1 = bf(st, rt)
    pos = bf_1(st, rt)
    return pos


if __name__ == "__main__":
    st = "Ilovezhaypqngfei.coldlar.cc we love bitcoin"
    rt = "coin"
    print(main(st, rt))

File: string/BF.c
#include "stdio.h"
#include "string.h"

#define SIZE 100

//字符串的模式匹配
int bf(char* s, char* t, int* a){
    int i=0, j=0;
    int slen = strlen(s);
    int tlen = strlen(t);
    while (i< slen && j<tlen)
    {
        printf("i=%d, j=%d\n", i, j);
        if (s[i] == t[j]){
            i++;
            j++;
        }
        else{
            i = i - j + 1;   // 失配时,i指针的回溯位置为本次匹配起始位置的下一个位置
            j = 0;
        }
    }
    printf("i=%d, j=%d\n", i, j);
    if (j == tlen)     // 当j指针完全遍历模式串,说明匹配完成
    {
        *a = (i - j);
        return 1;
    }
    return 0;
}

int main(){
    int pos=0;
    // char* st = "Ilovecold.cc";
    // char* rt = "vec";

    char* st = "Ilovezhaypqngfei.cababaaabaoldlar.cc we love bitcoin";
    char* rt = "ababaaaba";
    
    bf(st, rt, &pos);
    printf("子串的位置为: %d\n", pos);
}